### PR TITLE
Fix Issue 23900 - @safe is allowed in inline asm

### PR DESCRIPTION
**Pull request by ntrel**

### CODE REVIEW

File: compiler/src/dmd/statementsem.d [hunk 1]
```diff
@@ -3593,7 +3593,9 @@ Statement statementSemanticVisit(Statement s, Scope* sc)
             cas.error("`asm` statement is assumed to be impure - mark it with `pure` if it is not");
         if (!(cas.stc & STC.nogc) && sc.func.setGC(cas.loc, "`asm` statement in %s `%s` is assumed to use the GC - mark it with `@nogc` if it does not"))
             cas.error("`asm` statement is assumed to use the GC - mark it with `@nogc` if it does not");
-        if (!(cas.stc & (STC.trusted | STC.safe)))
+        if (cas.stc & STC.safe)
+            cas.error("`asm` statement cannot be marked `@safe`, use `@system` or `@trusted` instead");
```
Comment (maxhaton) [inline compiler/src/dmd/statementsem.d:3597]: Deprecation first

**Comment by Dany12LU** [comment]
@ntrel Check this https://github.com/dlang/dmd/blob/54b6a2f2b0a129fa3c4a994658d9f99e3a6d4ee3/compiler/test/compilable/test12979a.d#L4, we should remove all this

**Review by dkorpel — APPROVED** [review]

**Comment by Dany12LU** [comment]
Do we really have to wait for D2.114 to deprecate this? This is not a feature by any means, this is a mistake of whoever implemented those checks from the beginning

**Comment by ntrel** [comment]
> This is not a feature by any means, this is a mistake of whoever implemented those checks from the beginning

Well they did write a test for `asm @safe`, so perhaps it was intentional. Either way people could be using it assuming the compiler can infer whether the instructions are safe or not.

**Comment by dkorpel** [comment]
> Do we really have to wait for D2.114 to deprecate this? This is not a feature by any means, this is a mistake of whoever implemented those checks from the beginning

It's a deprecation starting with the next release. At 2.114 is will become an error.

**Comment by Dany12LU** [comment]
This deprecation maybe should be in the changelog?

**Comment by schveiguy** [comment]
> This deprecation maybe should be in the changelog?

Closing of the bug will be in the changelog. It gets added automatically.

**Pull request merged**